src/libcore/iter/adapters/fuse.rs

```diff
@@ -173,35 +163,135 @@ impl<I> ExactSizeIterator for Fuse<I>
 where
     I: ExactSizeIterator,
 {
-    default fn len(&self) -> usize {
+    fn len(&self) -> usize {
+        FuseExactSizeIteratorImpl::len(self)
+    }
+
+    fn is_empty(&self) -> bool {
+        FuseExactSizeIteratorImpl::is_empty(self)
+    }
+}
+
+unsafe impl<I> TrustedRandomAccess for Fuse<I>
+where
+    I: TrustedRandomAccess,
+{
+    unsafe fn get_unchecked(&mut self, i: usize) -> I::Item {
         match self.iter {
-            Some(ref iter) => iter.len(),
+            Some(ref mut iter) => iter.get_unchecked(i),
+            // SAFETY: the caller asserts there is an item at `i`, so we're not exhausted.
+            None => intrinsics::unreachable(),
+        }
+    }
+
+    fn may_have_side_effect() -> bool {
+        I::may_have_side_effect()
+    }
+}
+
+// Fuse specialization trait
+// Iterators and DoubleEndedIterators cannot be overlapped successfully
+// So, they're separated into each it's own trait to provide internal implementations
+// Similarly, ExactSizeIterators cannot be overlapped, so requires its own trait
```

Comment:
I think you _could_ bundle them all in one trait, if you add extra constraints on individual methods as needed. You can see an example of this in `ZipImpl`:
https://github.com/rust-lang/rust/blob/edc02580e4e80476ac1ded2cc1008eaf8b8400e6/src/libcore/iter/adapters/zip.rs#L83-L86

But I think it's also fine to have them in separate specialization traits.

Reply:
Okay I think that might help the optimizer inline that entire thing away. I have no facts to back my claim, of course. Just going by intuition. I'll try that out and see how it performs.

Reply:
I don't think it would matter for inlining -- the initial call depth is the same either way.

I think the code is fine your way, or we could consolidate like `ZipImpl` does. If nothing else, it would read a little easier, perhaps. You would have one `impl FuseImpl` with defaults, and just one specialized `impl FuseImpl` for `I: FusedIterator`. You don't need further specializations, because the extra double-ended/exact constraints are only needed on the specific methods.

```rust
// reduced example:
trait FuseImpl<I> {
    type Item;
    fn next(&mut self) -> Option<Self::Item>;
    fn next_back(&mut self) -> Option<Self::Item> where I: DoubleEndedIterator;
    fn len(&self) -> usize where I: ExactSizeIterator;
}
impl<I: Iterator> FuseImpl<I> for Fuse<I> {
    type Item = I::Item;
    default fn next(&mut self) -> Option<Self::Item> { ... }
    default fn next_back(&mut self) -> Option<Self::Item> where I: DoubleEndedIterator { ... }
    default fn len(&self) -> usize where I: ExactSizeIterator { ... }
}
impl<I: FusedIterator> FuseImpl<I> for Fuse<I> {
    fn next(&mut self) -> Option<Self::Item> { ... }
    fn next_back(&mut self) -> Option<Self::Item> where I: DoubleEndedIterator { ... }
    fn len(&self) -> usize where I: ExactSizeIterator { ... }
}
```

We should clarify this before merging though, because I don't think this "cannot be overlapped" comment is correct.